Issue with snapshot using delta on spark-dbt

### Describe the bug
Hi All,
Im creating snapshot using Delta using local spark. But after the Initial snapshot, from the second dbt snapshot command, im getting the error snapshot `target is not a snapshot table (missing "dbt_scd_id", "dbt_valid_from", "dbt_valid_to")`.
Im guessing this is because the select statement doesnt contain these fields as these are generated by dbt. But is there something Im missing or do we need to add these to select statement manually  from second run ? Any help would be great !

### Steps To Reproduce
1) Connect dbt to Spark Thrift server
2) create a snapshot query using delta with followiing configs :
     file_format=delta
     unique_key= your unique key
     strategy= timestamp
     updated_at = your timestamp column

    

### Expected behavior
scd2 operation on all `dbt snapshot`  queries

### Screenshots and log output
 `target is not a snapshot table (missing "dbt_scd_id", "dbt_valid_from", "dbt_valid_to")` from second `dbt snapshot` command

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [ ] bigquery
- [ ] snowflake
- [x] other (specify:  spark ( local ))


**The output of `dbt --version`:**
```
0.20.0
```

**The operating system you're using:**
RHEL 7 , Ubuntu etc
**The output of `python --version`:**
python3.9.6, python3.8
### Additional context
Add any other context about the problem here.

Hi @jtcohen6 , Following is the create table command form logs : 
```
create or replace table snapshots_new_1.my_fun_snapshot
    
    
    using delta
    
    
    
    
    
    as
      

    select *,
        md5(coalesce(cast(id as string ), '')
         || '|' || coalesce(cast(updated_at as string ), '')
        ) as dbt_scd_id,
        updated_at as dbt_updated_at,
        updated_at as dbt_valid_from,
        nullif(updated_at, updated_at) as dbt_valid_to
    from (
        
    

    SELECT
        1 AS id,
        CURRENT_TIMESTAMP AS updated_at
    ) sbq
```

Also , Im testing this both in my local WSL with python3.8 as well as in a organization provided RHEL7 system with python3.9

in WSL : 
Following error also shows up in the log recently : 
```
/* {"app": "dbt", "dbt_version": "0.20.0", "profile_name": "default", "target_name": "dev", "connection_name": "list_None_curated_source_dbt_test__audit"} */
show table extended in curated_source_dbt_test__audit like '*'
  
2021-08-03 06:48:23.076050 (ThreadPoolExecutor-1_0): Database Error
  org.apache.hive.service.cli.HiveSQLException: Error running query: org.apache.spark.sql.catalyst.analysis.NoSuchDatabaseException: Database 'curated_source_dbt_test__audit' not found
  	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.org$apache$spark$sql$hive$thriftserver$SparkExecuteStatementOperation$$execute(SparkExecuteStatementOperation.scala:361)
  	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.$anonfun$run$2(SparkExecuteStatementOperation.scala:263)
  	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
  	at org.apache.spark.sql.hive.thriftserver.SparkOperation.withLocalProperties(SparkOperation.scala:78)
  	at org.apache.spark.sql.hive.thriftserver.SparkOperation.withLocalProperties$(SparkOperation.scala:62)
  	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.withLocalProperties(SparkExecuteStatementOperation.scala:43)
  	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.run(SparkExecuteStatementOperation.scala:263)
  	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.run(SparkExecuteStatementOperation.scala:258)
  	at java.base/java.security.AccessController.doPrivileged(Native Method)
  	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
  	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1730)
  	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2.run(SparkExecuteStatementOperation.scala:272)
  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  	at java.base/java.lang.Thread.run(Thread.java:829)
  Caused by: org.apache.spark.sql.catalyst.analysis.NoSuchDatabaseException: Database 'curated_source_dbt_test__audit' not found
  	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.requireDbExists(SessionCatalog.scala:192)
  	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.listTables(SessionCatalog.scala:951)
  	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.listTables(SessionCatalog.scala:932)
  	at org.apache.spark.sql.execution.command.ShowTablesCommand.$anonfun$run$42(tables.scala:868)
  	at scala.Option.map(Option.scala:230)
  	at org.apache.spark.sql.execution.command.ShowTablesCommand.run(tables.scala:868)
  	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
  	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
  	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
  	at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:228)
  	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3687)
  	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
  	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
  	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
  	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
  	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3685)
  	at org.apache.spark.sql.Dataset.<init>(Dataset.scala:228)
  	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:99)
  	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:96)
  	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:615)
  	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:610)
  	at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:650)
  	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.org$apache$spark$sql$hive$thriftserver$SparkExecuteStatementOperation$$execute(SparkExecuteStatementOperation.scala:325)
  	... 16 more
  
2021-08-03 06:48:23.076894 (ThreadPoolExecutor-1_0): Error while running:
macro list_relations_without_caching
2021-08-03 06:48:23.077663 (ThreadPoolExecutor-1_0): Runtime Error
  Database Error
    org.apache.hive.service.cli.HiveSQLException: Error running query: org.apache.spark.sql.catalyst.analysis.NoSuchDatabaseException: Database 'curated_source_dbt_test__audit' not found
    	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.org$apache$spark$sql$hive$thriftserver$SparkExecuteStatementOperation$$execute(SparkExecuteStatementOperation.scala:361)
    	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.$anonfun$run$2(SparkExecuteStatementOperation.scala:263)
    	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
    	at org.apache.spark.sql.hive.thriftserver.SparkOperation.withLocalProperties(SparkOperation.scala:78)
    	at org.apache.spark.sql.hive.thriftserver.SparkOperation.withLocalProperties$(SparkOperation.scala:62)
    	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.withLocalProperties(SparkExecuteStatementOperation.scala:43)
    	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.run(SparkExecuteStatementOperation.scala:263)
    	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.run(SparkExecuteStatementOperation.scala:258)
    	at java.base/java.security.AccessController.doPrivileged(Native Method)
    	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
    	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1730)
    	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2.run(SparkExecuteStatementOperation.scala:272)
    	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    	at java.base/java.lang.Thread.run(Thread.java:829)
    Caused by: org.apache.spark.sql.catalyst.analysis.NoSuchDatabaseException: Database 'curated_source_dbt_test__audit' not found
    	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.requireDbExists(SessionCatalog.scala:192)
    	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.listTables(SessionCatalog.scala:951)
    	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.listTables(SessionCatalog.scala:932)
    	at org.apache.spark.sql.execution.command.ShowTablesCommand.$anonfun$run$42(tables.scala:868)
    	at scala.Option.map(Option.scala:230)
    	at org.apache.spark.sql.execution.command.ShowTablesCommand.run(tables.scala:868)
    	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
    	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
    	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
    	at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:228)
    	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3687)
    	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
    	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
    	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
    	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
    	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
    	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3685)
    	at org.apache.spark.sql.Dataset.<init>(Dataset.scala:228)
    	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:99)
    	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
    	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:96)
    	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:615)
    	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
    	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:610)
    	at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:650)
    	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.org$apache$spark$sql$hive$thriftserver$SparkExecuteStatementOperation$$execute(SparkExecuteStatementOperation.scala:325)
    	... 16 more

```

`curated_source` is the default schema i had provided in the `profiles.yml` in `.dbt` folder. But i have provided custom schema for the snapshot also . Following is my `snapshot sql` : 
```
{% snapshot my_fun_snapshot %}
    {{ config(
        target_schema = 'snapshots_new_1',
        unique_key = 'id',
        file_format = 'delta',
        strategy = 'timestamp',
        updated_at = 'updated_at',
    ) }}

    SELECT
        1 AS id,
        CURRENT_TIMESTAMP AS updated_at
{% endsnapshot %}
```
This i have done on a fresh dbt project with the snapshot sql alone in the `snapshots` folder


In RHEL7:

Im not getting any errors like the above in the logs .. just the final Compilation error message

Im using a docker based setup to start up following services : 
 - minio
 - mariadb ( external metastore )
 - Hive metastore standalone

my `spark-defaults.conf` : 
```
spark.sql.extensions             io.delta.sql.DeltaSparkSessionExtension
spark.sql.catalog.spark_catalog  org.apache.spark.sql.delta.catalog.DeltaCatalog
hive.metastore.uris              thrift://0.0.0.0:9083/
spark.sql.warehouse.dir          s3a://spark/warehouse/
spark.hadoop.fs.s3a.endpoint     http://127.0.0.1:9000
spark.hadoop.fs.s3a.impl         org.apache.hadoop.fs.s3a.S3AFileSystem
spark.hadoop.fs.s3a.access.key   accesskey
spark.hadoop.fs.s3a.secret.key   secretkey
spark.driver.extraClassPath      /opt/spark/lib/*
spark.executor.extraClassPath    /opt/spark/lib/*
spark.executor.cores             3
spark.driver.cores               2
```

Using this I start my `$SPARK_HOME/sbin/start-thriftserver.sh`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,11 +1,13 @@
 ## dbt-spark 0.21.0 (Release TBD)
 
 ### Fixes
+- Enhanced get_columns_in_relation method to handle a bug in open source deltalake which doesnt return schema details in `show table extended in databasename like '*'` query output. This impacts dbt snapshots if file format is open source deltalake ([#207](https://github.com/dbt-labs/dbt-spark/pull/207))
 - Add pyodbc import error message to dbt.exceptions.RuntimeException to get more detailed information when running `dbt debug` ([#192](https://github.com/dbt-labs/dbt-spark/pull/192))
 - Add support for ODBC Server Side Parameters, allowing options that need to be set with the `SET` statement to be used ([#201](https://github.com/dbt-labs/dbt-spark/pull/201))
 - Add `retry_all` configuration setting to retry all connection issues, not just when the `_is_retryable_error` function determines ([#194](https://github.com/dbt-labs/dbt-spark/pull/194))
 
 ### Contributors
+- [@harryharanb](https://github.com/harryharanb) ([#207](https://github.com/dbt-labs/dbt-spark/pull/207))
 - [@JCZuurmond](https://github.com/JCZuurmond) ([#192](https://github.com/fishtown-analytics/dbt-spark/pull/192))
 - [@jethron](https://github.com/jethron) ([#201](https://github.com/fishtown-analytics/dbt-spark/pull/201))
 - [@gregingenii](https://github.com/gregingenii) ([#194](https://github.com/dbt-labs/dbt-spark/pull/194))

--- a/dbt/adapters/spark/impl.py
+++ b/dbt/adapters/spark/impl.py
@@ -212,11 +212,15 @@ class SparkAdapter(SQLAdapter):
                                 for cached_relation in cached_relations
                                 if str(cached_relation) == str(relation)),
                                None)
-        if cached_relation is None or cached_relation.information is None:
+        columns = []
+        if cached_relation and cached_relation.information:
+            columns = self.parse_columns_from_information(cached_relation)
+        if not columns:
+            # in open source delta 'show table extended' query output doesnt return relation's schema
+            # if columns are empty from cache, use get_columns_in_relation spark macro
+            # which would execute 'describe extended tablename' query
             rows: List[agate.Row] = super().get_columns_in_relation(relation)
             columns = self.parse_describe_extended(relation, rows)
-        else:
-            columns = self.parse_columns_from_information(cached_relation)
         return columns
 
     def parse_columns_from_information(

--- a/dbt/adapters/spark/impl.py
+++ b/dbt/adapters/spark/impl.py
@@ -216,8 +216,9 @@ class SparkAdapter(SQLAdapter):
         if cached_relation and cached_relation.information:
             columns = self.parse_columns_from_information(cached_relation)
         if not columns:
-            # in open source delta 'show table extended' query output doesnt return relation's schema
-            # if columns are empty from cache, use get_columns_in_relation spark macro
+            # in open source delta 'show table extended' query output doesnt
+            # return relation's schema. if columns are empty from cache,
+            # use get_columns_in_relation spark macro
             # which would execute 'describe extended tablename' query
             rows: List[agate.Row] = super().get_columns_in_relation(relation)
             columns = self.parse_describe_extended(relation, rows)

